README.md
RKit
====

Responsive kit for front-end developers


changelog.txt
12/26/14
---------
- RKit:imageLoader component
IE bag with caching images

- Replace Grunt task to Gulp

07/07/14
---------
Release v1.0.0


Gulpfile.js
var gulp = require('gulp');
var notify = require('gulp-notify');
var concat = require('gulp-concat');
var uglify = require('gulp-uglify');
var rename = require('gulp-rename');
var header = require('gulp-header');
var pkg = require('./package.json');

/**
 * Base Configuration
 * @type {{src: string}}
 */
var js = {
    'buildName' : 'rkit.js',
    'suffix' : '.min',
    'dest' : 'js/dest',
    'src' : [
        'js/src/core.js',
        'js/src/utility.js',
        'js/src/yoga.js',
        'js/src/media.js',
        'js/src/app.js'
    ],
    'banner' : [
        '/**',
        ' * ${name} - ${description}',
        ' * @version ${version}',
        ' * @date ${date}',
        ' */ ',
        '\r\n'
    ].join('\n')
};

var bannerData = {
    'name' : pkg.name,
    'version' : pkg.version,
    'description' : pkg.description,
    'date' : new Date()
};

/**
 * Describe tasks
 */
gulp.task('js', function() {
   gulp.src(js.src)
       .pipe(concat(js.buildName))
       .pipe(header(js.banner, bannerData))
       .pipe(gulp.dest(js.dest))
       .pipe(rename({
           'suffix' : js.suffix
       }))
       .pipe(uglify())
       .pipe(header(js.banner, bannerData))
       .pipe(gulp.dest(js.dest))
       .pipe(notify('Js files were concatenated and compressed'));
});

/**
 * Watching for files
 */
gulp.watch(js.src, ['js']);

/**
 * Default task
 */
gulp.task('default', ['js']);


Gruntfile.js
module.exports = function(grunt) {

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        concat: {
            dist: {
                src: [
                    'js/src/core.js',
                    'js/src/utility.js',
                    'js/src/yoga.js',
                    'js/src/media.js',
                    'js/src/app.js'
                ],
                dest: 'js/dest/<%= pkg.name %>.js'
            }
        },
        jshint: {
            files: ['Gruntfile.js', 'js/**/*.js', 'test/**/*.js'],
            options: {
                // options here to override JSHint defaults
                globals: {
                    jQuery: true,
                    console: true,
                    module: true,
                    document: true
                }
            }
        },
        uglify: {
            options: {
                mangle: false,
                banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' +
                    '<%= grunt.template.today("yyyy-mm-dd") %> */\n'
            },
            my_target: {
                files: {
                    'js/dest/<%= pkg.name %>.min.js': ['js/dest/<%= pkg.name %>.js']
                }
            }
        },
        watch: {
            files: ['<%= jshint.files %>'],
            tasks: ['concat', 'uglify']
        }
    });

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-qunit');

    grunt.registerTask('test', ['jshint']);
    grunt.registerTask('min', ['uglify']);
    grunt.registerTask('default', ['concat', 'watch']);

};


js/src/yoga.js
/**
 * RKit: yoga
 * smart block move
 *
 * Used data attributes
 * data-push="{selector}"
 * data-dir="{before|after|append|prepend}"
 * data-bp="{media{M|T|D}}"
 */
RKit.Yoga = (function($) {
   'use strict';

    var base, config;

    /* default config */
    config = {
        className: 'js-push',
        tempClass: 'js-rkit-temp'
    };

    base = {};

    /**
     * Blocks storage
     * @type {Array}
     */
    base.blocksStorage = [];

    /**
     * Move block
     * @param media
     */
    base.move = function(media) {
        var i, len, pushTo, clone;
        len = base.blocksStorage.length;
        for (i = 0; i < len; i++) {
            if ( base.blocksStorage[i].pushBreakPoint == media && !base.blocksStorage[i].isInsert ) {
                pushTo = $(base.blocksStorage[i].pushTo);
                clone = base.blocksStorage[i].$block.clone().addClass(config.tempClass + '-' + i);

                switch (base.blocksStorage[i].pushDir) {
                    case 'before':
                        pushTo.before(clone);
                        break;
                    case 'after':
                        pushTo.after(clone);
                        break;
                    case 'append':
                        pushTo.append(clone);
                        break;
                    case 'prepend':
                        pushTo.prepend(clone);
                        break;
                }
                base.blocksStorage[i].isInsert = true;
                base.blocksStorage[i].$block.hide();
            } else if ( base.blocksStorage[i].pushBreakPoint == media && base.blocksStorage[i].isInsert ) {
                base.blocksStorage[i].$block.hide();
                $('.' + config.tempClass + '-' + i).show();
            } else if ( base.blocksStorage[i].pushBreakPoint !== media && base.blocksStorage[i].isInsert ) {
                base.blocksStorage[i].$block.show();
                $('.' + config.tempClass + '-' + i).hide();
            }
        }
    };

    /**
     * Cache all blocks
     */
    base.getBlocks = function() {
        var block, pushTo, pushDir, pushBreakPoint;
        $('.' + config.className).each(function() {
            block = $(this);
            pushTo = block.data('push');
            pushDir = block.data('dir');
            pushBreakPoint = block.data('bp');

            base.blocksStorage.push({
                $block: block,
                pushTo: pushTo,
                pushDir: pushDir,
                pushBreakPoint: pushBreakPoint,
                isInsert: false
            });
        });
    };

    /**
     * Initialize
     */
    base.init = function(conf) {
        if (typeof conf === "object") {
            config = $.extend({}, config, conf);
        }
        base.getBlocks();
    };

    return base;
})(jQuery);

RKit.Y = RKit.Yoga;


js/src/media.js
/**
 * RKit media
 */
RKit.Media = (function($) {
    'use strict';

    var api, config, Win, WinW;

    /* default config */
    config = {
        'breakpoints': [
            ['mediaM', 768],
            ['mediaT', [768, 1024]],
            ['mediaD', 1024]
        ]
    };

    /* stash public methods */
    api = {};

    /**
     * Reset all media flags
     */
    api.resetMedia = function() {
        for(var i = 0, mediaLength = config.breakpoints.length; i < mediaLength; i++) {
            window[config.breakpoints[i][0]] = false;
        }
    };

    /**
     * Set media break point
     * @param media
     */
    api.setMedia = function(media) {
        /* reset all media */
        api.resetMedia();

        /* set media */
        window[media] = true;

        /* trigger event */
        Win.trigger(media);

        /* RKit Module: Yoga */
        if ( typeof RKit.Y === "object" ) {
            RKit.Y.move(media);
        }
    };
    /**
     * Get media break point
     */
    api.getMedia = function() {
        WinW = RKit.getViewPortWidth();
        /* mobile */
        if ( !(window[ config.breakpoints[0][0] ]) && (WinW <= config.breakpoints[0][1] - 1) ) {
            api.setMedia(config.breakpoints[0][0]);
        }

        /* tablet */
        if ( !(window[ config.breakpoints[1][0] ]) && (WinW >= config.breakpoints[1][1] && WinW <= config.breakpoints[1][2]) ) {
            api.setMedia(config.breakpoints[1][0]);
        }

        /* desktop */
        if ( !(window[ config.breakpoints[2][0] ]) && WinW > config.breakpoints[2][1]) {
            api.setMedia(config.breakpoints[2][0]);
        }

    };

    /**
     * Initialize
     */
    api.init = function(conf) {
        if (typeof conf === "object") {
            config = $.extend({}, config, conf);
        }

        Win = $(window);
        setTimeout(api.getMedia, 0);
        Win.resize(api.getMedia);
    };

    return api;
})(jQuery);


js/src/utility.js
/**
 * RKit Helper Methods
 */
(function($) {
    'use strict';

    window.RKit = RKit || {};

    RKit.getViewPortWidth = function() {
        var viewPortWidth;

        if (typeof window.innerWidth != 'undefined') {
            viewPortWidth  = window.innerWidth;
        }
        else if (typeof document.documentElement != 'undefined' && typeof document.documentElement.clientWidth != 'undefined' && document.documentElement.clientWidth != 0) {
            viewPortWidth  = document.documentElement.clientWidth;
        }
        else {
            viewPortWidth  = document.getElementsByTagName('body')[0].clientWidth;
        }

        return viewPortWidth;
    };

    /**
     * Detect mobile
     */
    RKit.isMobile = function() {
        var re = /(Android|webOS|iPhone|iPad|iPod|BlackBerry|Windows Phone)/i;
        return re.test(window.navigator.userAgent)
    };

    /**
     * Detect retina displays
     */
    RKit.isRetina = function() {
        if (window.devicePixelRatio) {
            return window.devicePixelRatio >= 2;
        } else {
            return false;
        }
    };

    /**
     * Image loader
     * @param {object} container - jquery object which contains images
     */
    RKit.imageLoader = function(container, fn) {
        var target, n, c, images, image;

        if (typeof container === "object") {
            target = container;
        } else if (typeof container === "string") {
            target = $(container);
        }

        images = target.find('img');
        n = images.length;
        c = 0;
        images.each(function() {
            image = $(this);
            image.one('load error', function() {
                c++;
                if (c === n) {
                    if (typeof fn === "function") {
                        fn.call(this);
                    }
                }
            });
        });
    };

})(jQuery);


js/dest/rkit.js
/**
 * RKit - Responsive Kit which contains usefull components for develop reponsive web application.
 * @version 1.0.0
 * @date Fri Dec 26 2014 13:07:52 GMT+0100 (Central Europe Standard Time)
 */ 

/**
 * RKit - responsive web kit for developers
 * Core
 */
var RKit = {
    version: '1.0.0'
};

/**
 * RKit Helper Methods
 */
(function($) {
    'use strict';

    window.RKit = RKit || {};

    RKit.getViewPortWidth = function() {
        var viewPortWidth;

        if (typeof window.innerWidth != 'undefined') {
            viewPortWidth  = window.innerWidth;
        }
        else if (typeof document.documentElement != 'undefined' && typeof document.documentElement.clientWidth != 'undefined' && document.documentElement.clientWidth != 0) {
            viewPortWidth  = document.documentElement.clientWidth;
        }
        else {
            viewPortWidth  = document.getElementsByTagName('body')[0].clientWidth;
        }

        return viewPortWidth;
    };

    /**
     * Detect mobile
     */
    RKit.isMobile = function() {
        var re = /(Android|webOS|iPhone|iPad|iPod|BlackBerry|Windows Phone)/i;
        return re.test(window.navigator.userAgent)
    };

    /**
     * Detect retina displays
     */
    RKit.isRetina = function() {
        if (window.devicePixelRatio) {
            return window.devicePixelRatio >= 2;
        } else {
            return false;
        }
    };

    /**
     * Image loader
     * @param {object} container - jquery object which contains images
     */
    RKit.imageLoader = function(container, fn) {
        var target, n, c, images, image, imageIterator;

        if (typeof container === "object") {
            target = container;
        } else if (typeof container === "string") {
            target = $(container);
        }

        images = target.find('img');
        n = images.length;
        c = 0;
        imageIterator = function() {
            c++;
            if (c === n) {
                if (typeof fn === "function") {
                    fn.call(this);
                }
            }
        };

        images.each(function() {
            image = $(this);
            /* if image was cached */
            if (image[0].complete) {
                imageIterator();
            } else {
                image.one('load error', function() {
                    imageIterator();
                });
            }
        });
    };

})(jQuery);

/**
 * RKit: yoga
 * move block from one place to another with reverting when breakpoint was leave
 *
 * Used data attributes
 * data-push="{selector}"
 * data-dir="{after|before|append|prepend}"
 * data-bp="{media{M|T|D}}"
 */
RKit.Yoga = (function($) {
   'use strict';

    var api, config;

    /* default config */
    config = {
        dataPush: "push"
    };

    api = {};

    /**
     * Blocks storage
     * @type {Array}
     */
    api.blocksStorage = [];

    /**
     * Cache all blocks
     */
    api.getBlocks = function() {
        var $block, $pushTo, pushDir, pushBreakPoint, $startBlockPos, startDir;
        $("[data-" + config.dataPush +"]").each(function() {
            $block = $(this);
            $pushTo = $($block.data(config.dataPush));
            pushDir = $block.data('dir');
            pushBreakPoint = $block.data('bp');

            /* find initial block position */
            if ($block.prev().length) {
                $startBlockPos = $block.prev();
                startDir = "after";
            } else if ($block.next().length) {
                $startBlockPos = $block.next();
                startDir = "before";
            } else {
                $startBlockPos = $block.parent();
                startDir = "append";
            }

            api.blocksStorage.push({
                $block: $block,
                $pushTo: $pushTo,
                pushDir: pushDir,
                pushBreakPoint: pushBreakPoint,
                $startBlockPos: $startBlockPos,
                startDir: startDir,
                isInsert: false
            });
        });
    };

    /**
     * Move block
     * @param media
     */
    api.move = function(media) {
        var i, len, $pushTo, $block, isInsert, breakPoint, dir, $startBlockPos, startDir;
        len = api.blocksStorage.length;
        for (i = 0; i < len; i++) {

            $pushTo = api.blocksStorage[i].$pushTo;
            $block = api.blocksStorage[i].$block;
            breakPoint = api.blocksStorage[i].pushBreakPoint;
            dir = api.blocksStorage[i].pushDir;
            $startBlockPos = api.blocksStorage[i].$startBlockPos;
            startDir = api.blocksStorage[i].startDir;
            isInsert = api.blocksStorage[i].isInsert;

            if ( breakPoint == media && !isInsert ) {
                switch (dir) {
                    case 'before':
                        $block.insertBefore($pushTo);
                        break;
                    case 'after':
                        $block.insertAfter($pushTo);
                        break;
                    case 'append':
                        $block.appendTo($pushTo);
                        break;
                    case 'prepend':
                        $block.prependTo($pushTo);
                        break;
                }
                api.blocksStorage[i].isInsert = true;
            } else if ( breakPoint == media && isInsert ) {
                switch (dir) {
                    case 'before':
                        $block.insertBefore($pushTo);
                        break;
                    case 'after':
                        $block.insertAfter($pushTo);
                        break;
                    case 'append':
                        $block.appendTo($pushTo);
                        break;
                    case 'prepend':
                        $block.prependTo($pushTo);
                        break;
                }
            } else if ( breakPoint !== media && isInsert ) {
                switch (startDir) {
                    case 'before':
                        $block.insertBefore($startBlockPos);
                        break;
                    case 'after':
                        $block.insertAfter($startBlockPos);
                        break;
                    case 'append':
                        $block.appendTo($startBlockPos);
                        break;
                    case 'prepend':
                        $block.prependTo($startBlockPos);
                        break;
                }
            }
        }
    };

    /**
     * Initialize
     */
    api.init = function() {
        api.getBlocks();
    };

    return api;
})(jQuery);

RKit.Y = RKit.Yoga;

/**
 * RKit media
 */
RKit.Media = (function($) {
    'use strict';

    var api, config, Win, WinW;

    /* default config */
    config = {
        'breakpoints': [
            ['mediaM', 768],
            ['mediaT', [768, 1024]],
            ['mediaD', 1024]
        ]
    };

    /* stash public methods */
    api = {};

    /**
     * Reset all media flags
     */
    api.resetMedia = function() {
        for(var i = 0, mediaLength = config.breakpoints.length; i < mediaLength; i++) {
            window[config.breakpoints[i][0]] = false;
        }
    };

    /**
     * Set media break point
     * @param media
     */
    api.setMedia = function(media) {
        /* reset all media */
        api.resetMedia();

        /* set media */
        window[media] = true;

        /* trigger event */
        Win.trigger(media);

        /* RKit Module: Yoga */
        if ( typeof RKit.Y === "object" ) {
            RKit.Y.move(media);
        }
    };
    /**
     * Get media break point
     */
    api.getMedia = function() {
        WinW = RKit.getViewPortWidth();
        /* mobile */
        if ( !(window[ config.breakpoints[0][0] ]) && (WinW <= config.breakpoints[0][1] - 1) ) {
            api.setMedia(config.breakpoints[0][0]);
        }

        /* tablet */
        if ( !(window[ config.breakpoints[1][0] ]) && (WinW >= config.breakpoints[1][1] && WinW <= config.breakpoints[1][2]) ) {
            api.setMedia(config.breakpoints[1][0]);
        }

        /* desktop */
        if ( !(window[ config.breakpoints[2][0] ]) && WinW > config.breakpoints[2][1]) {
            api.setMedia(config.breakpoints[2][0]);
        }

    };

    /**
     * Initialize
     */
    api.init = function(conf) {
        if (typeof conf === "object") {
            config = $.extend({}, config, conf);
        }

        Win = $(window);
        setTimeout(api.getMedia, 0);
        Win.resize(api.getMedia);
    };

    return api;
})(jQuery);

/**
 * RKit: Application Initialize
 *
 */
RKit.init = function (conf) {
    /* Module: Media */
    if (typeof conf.media === "object") {
        RKit.Media.init(conf.media);
    }

    /* Module: Yoga */
    if (conf.yoga) {
        RKit.Y.init();
    }
};
